ext/mixed/js/display-generator.js

```diff
@@ -346,4 +481,85 @@ class DisplayGenerator {
         }
     }
+
+    static _getPitchInfos(definition) {
+        const results = new Map();
+
+        const expressions = definition.expressions;
+        const sources = Array.isArray(expressions) ? expressions : [definition];
+        for (const {pitches: expressionPitches, expression} of sources) {
+            for (const {reading, pitches, dictionary} of expressionPitches) {
+                let dictionaryResults = results.get(dictionary);
+                if (typeof dictionaryResults === 'undefined') {
+                    dictionaryResults = [];
+                    results.set(dictionary, dictionaryResults);
+                }
+
+                for (const {position, tags} of pitches) {
+                    let pitchInfo = DisplayGenerator._findExistingPitchInfo(reading, position, tags, dictionaryResults);
+                    if (pitchInfo === null) {
+                        pitchInfo = {expressions: new Set(), reading, position, tags};
+                        dictionaryResults.push(pitchInfo);
+                    }
+                    pitchInfo.expressions.add(expression);
+                }
+            }
+        }
+
+        return [...results.entries()];
+    }
+
+    static _findExistingPitchInfo(reading, position, tags, pitchInfoList) {
+        for (const pitchInfo of pitchInfoList) {
+            if (
+                pitchInfo.reading === reading &&
+                pitchInfo.position === position &&
+                DisplayGenerator._areTagListsEqual(pitchInfo.tags, tags)
+            ) {
+                return pitchInfo;
+            }
+        }
+        return null;
+    }
+
+    static _areTagListsEqual(tagList1, tagList2) {
+        const ii = tagList1.length;
+        if (tagList2.length !== ii) { return false; }
+
+        for (let i = 0; i < ii; ++i) {
+            const tag1 = tagList1[i];
+            const tag2 = tagList2[i];
+            if (tag1.name !== tag2.name || tag1.dictionary !== tag2.dictionary) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    static _jpGetKanaMorae(text) {
+        // This function splits Japanese kana reading into its individual mora
+        // components. It is assumed that the text is well-formed.
+        const smallKanaSet = DisplayGenerator._smallKanaSet;
+        const morae = [];
+        let i;
+        for (const c of text) {
+            if (smallKanaSet.has(c) && (i = morae.length) > 0) {
+                morae[i - 1] += c;
+            } else {
+                morae.push(c);
+            }
+        }
+        return morae;
+    }
+
+    static _jpCreateSmallKanaSet() {
+        return new Set(Array.from('ぁぃぅぇぉゃゅょゎァィゥェォャュョヮ'));
+    }
+
+    static _jpIsMoraPitchHigh(moraIndex, pitchAccentPosition) {
+        return pitchAccentPosition === 0 ? (moraIndex > 0) : (moraIndex < pitchAccentPosition);
+    }
 }
+
+DisplayGenerator._smallKanaSet = DisplayGenerator._jpCreateSmallKanaSet();
```

Comment:
There are now even more Japanese functions in `DisplayGenerator`. I remember that we let `_isCharacterKanji` slide because it was so simple, but this is more complex. Might be worth it to include `japanese.js` as a dependency now.

Also, does adding `_smallKanaSet` to the class have the same issues as callbacks (https://github.com/FooSoft/yomichan/issues/359)?

Reply:
> There are now even more Japanese functions in `DisplayGenerator`. I remember that we let `_isCharacterKanji` slide because it was so simple, but this is more complex. Might be worth it to include `japanese.js` as a dependency now.

There should probably be a new japanese.js for the frontend now, yeah.

> Also, does adding `_smallKanaSet` to the class have the same issues as callbacks (#359)?

Not really an issue per se, but it could be updated to use the same convention. This change was initially made before that, so that's partially why this style was used.

Reply:
> There should probably be a new japanese.js for the frontend now, yeah.

That's cool too if there's not much overlap.

Reply:
Updated in 0d80fcdf86745da133e4510eeea809a4eeafe120